Tpetra: Improve CrsMatrix::sumInto benchmark

mhoemmen opened the pull request:
@trilinos/tpetra @tjfulle 

## Description

Improve `Tpetra::CrsMatrix::sumIntoLocalValues` benchmark.  Here are some `MPI_RELEASE` results with GCC 4.9.3 and OpenMP.  Notice that calling `resumeFill` after first `fillComplete` causes `sumIntoLocalValues` to slow down to times comparable to that of `sumIntoLocalValues` before first `fillComplete`.

```
lclNumInds: 10000
maxNumEntPerRow: 28
numToInsert: 27
numTrials: 100
resumeFill: false
================================================================================

                      TimeMonitor results over 1 processor

Timer Name                                       Global time (num calls)
--------------------------------------------------------------------------------
CrsMatrix constructor                            4.196e-05 (1)
CrsMatrix fillComplete                           0.00188 (1)
Kokkos sumInto                                   0.05328 (1)
Whole benchmark                                  1.874 (1)
createRowAndColMap                               0.0001009 (1)
insertLocalValues loop                           0.006014 (1)
sumIntoLocalValues: after first fillComplete     0.00402 (1)
sumIntoLocalValues: before first fillComplete    1.808 (1)
================================================================================

lclNumInds: 10000
maxNumEntPerRow: 28
numToInsert: 27
numTrials: 100
resumeFill: true
================================================================================

                      TimeMonitor results over 1 processor

Timer Name                                       Global time (num calls)
--------------------------------------------------------------------------------
CrsMatrix constructor                            5.007e-05 (1)
CrsMatrix fillComplete                           0.001874 (1)
CrsMatrix resumeFill                             4.053e-06 (1)
Kokkos sumInto                                   0.05336 (1)
Whole benchmark                                  3.731 (1)
createRowAndColMap                               9.203e-05 (1)
insertLocalValues loop                           0.006702 (1)
sumIntoLocalValues: after first fillComplete     1.812 (1)
sumIntoLocalValues: before first fillComplete    1.856 (1)
================================================================================
```


## Motivation and Context
<!--- Why is this change required?  What problem does it solve? -->

mhoemmen commented:
Here's what happens when I comment out the branch at the [beginning of `CrsMatrix::sumIntoLocalValuesImpl`](https://github.com/trilinos/Trilinos/blob/c80327ff3efb396b208cfc7e51fd1c527408a54e/packages/tpetra/core/src/Tpetra_CrsMatrix_def.hpp#L3250):

```
lclNumInds: 10000
maxNumEntPerRow: 28
numToInsert: 27
numTrials: 100
resumeFill: false
================================================================================

                      TimeMonitor results over 1 processor

Timer Name                                       Global time (num calls)
--------------------------------------------------------------------------------
CrsMatrix constructor                            4.005e-05 (1)
CrsMatrix fillComplete                           0.001868 (1)
Kokkos sumInto                                   0.05332 (1)
Whole benchmark                                  0.3816 (1)
createRowAndColMap                               8.202e-05 (1)
insertLocalValues loop                           0.006909 (1)
sumIntoLocalValues: after first fillComplete     0.004022 (1)
sumIntoLocalValues: before first fillComplete    0.3153 (1)
================================================================================

lclNumInds: 10000
maxNumEntPerRow: 28
numToInsert: 27
numTrials: 100
resumeFill: true
================================================================================

                      TimeMonitor results over 1 processor

Timer Name                                       Global time (num calls)
--------------------------------------------------------------------------------
CrsMatrix constructor                            4.196e-05 (1)
CrsMatrix fillComplete                           0.001872 (1)
CrsMatrix resumeFill                             4.053e-06 (1)
Kokkos sumInto                                   0.05336 (1)
Whole benchmark                                  0.6929 (1)
createRowAndColMap                               9.179e-05 (1)
insertLocalValues loop                           0.007241 (1)
sumIntoLocalValues: after first fillComplete     0.315 (1)
sumIntoLocalValues: before first fillComplete    0.3152 (1)
================================================================================

```

mhoemmen commented:
Here is what happens when I comment out everything after the initial branch (so that only the "new code" runs):
```
lclNumInds: 10000
maxNumEntPerRow: 28
numToInsert: 27
numTrials: 100
resumeFill: false
================================================================================

                      TimeMonitor results over 1 processor

Timer Name                                       Global time (num calls)
--------------------------------------------------------------------------------
CrsMatrix constructor                            4.196e-05 (1)
CrsMatrix fillComplete                           0.002085 (1)
Kokkos sumInto                                   0.05878 (1)
Whole benchmark                                  1.789 (1)
createRowAndColMap                               9.489e-05 (1)
insertLocalValues loop                           0.006216 (1)
sumIntoLocalValues: after first fillComplete     0.004405 (1)
sumIntoLocalValues: before first fillComplete    1.717 (1)
================================================================================

lclNumInds: 10000
maxNumEntPerRow: 28
numToInsert: 27
numTrials: 100
resumeFill: true
================================================================================

                      TimeMonitor results over 1 processor

Timer Name                                       Global time (num calls)
--------------------------------------------------------------------------------
CrsMatrix constructor                            4.101e-05 (1)
CrsMatrix fillComplete                           0.002006 (1)
CrsMatrix resumeFill                             5.007e-06 (1)
Kokkos sumInto                                   0.05879 (1)
Whole benchmark                                  3.537 (1)
createRowAndColMap                               9.394e-05 (1)
insertLocalValues loop                           0.00692 (1)
sumIntoLocalValues: after first fillComplete     1.733 (1)
sumIntoLocalValues: before first fillComplete    1.737 (1)
================================================================================
```

mhoemmen commented:
Just to make sure: I put a throw after the "new" branch, and the throw didn't throw.  Here are the results:
```
lclNumInds: 10000
maxNumEntPerRow: 28
numToInsert: 27
numTrials: 100
resumeFill: false
================================================================================

                      TimeMonitor results over 1 processor

Timer Name                                       Global time (num calls)
--------------------------------------------------------------------------------
CrsMatrix constructor                            3.791e-05 (1)
CrsMatrix fillComplete                           0.001879 (1)
Kokkos sumInto                                   0.05878 (1)
Whole benchmark                                  1.81 (1)
createRowAndColMap                               8.917e-05 (1)
insertLocalValues loop                           0.006087 (1)
sumIntoLocalValues: after first fillComplete     0.00402 (1)
sumIntoLocalValues: before first fillComplete    1.739 (1)
================================================================================

lclNumInds: 10000
maxNumEntPerRow: 28
numToInsert: 27
numTrials: 100
resumeFill: true
================================================================================

                      TimeMonitor results over 1 processor

Timer Name                                       Global time (num calls)
--------------------------------------------------------------------------------
CrsMatrix constructor                            3.695e-05 (1)
CrsMatrix fillComplete                           0.001875 (1)
CrsMatrix resumeFill                             4.053e-06 (1)
Kokkos sumInto                                   0.05878 (1)
Whole benchmark                                  3.556 (1)
createRowAndColMap                               8.297e-05 (1)
insertLocalValues loop                           0.00591 (1)
sumIntoLocalValues: after first fillComplete     1.744 (1)
sumIntoLocalValues: before first fillComplete    1.745 (1)
================================================================================
```

mhoemmen commented:
These latest results are with atomic updates turned off, btw.

tjfulle commented:
what do you mean by new code?  I thought the only branches were static vs dynamic profile?

mhoemmen commented:
"New code" means the initial "static profile" branch at the top of `sumIntoLocalValuesImpl`.

mhoemmen commented:
Anyway, don't worry about that for now :-) .  I did some new experiments.  I reverted `Tpetra::CrsMatrix` to stock Trilinos develop, and added cases in the benchmark for creating the CrsMatrix with a const CrsGraph, vs. creating it with row and column Maps and an upper bound.  That doesn't matter at all.  The only thing that matters is whether the matrix (not the graph) is fill complete.

```
lclNumInds: 10000
maxNumEntPerRow: 28
numToInsert: 27
numTrials: 100
resumeFill: false
=============================================================================================

                            TimeMonitor results over 1 processor

Timer Name                                                        Global time (num calls)
---------------------------------------------------------------------------------------------
(const graph) CrsMatrix fillComplete                              1.192e-05 (1)
(const graph) Kokkos sumInto                                      0.0526 (1)
(const graph) sumIntoLocalValues: after first fillComplete        0.004022 (1)
(nonconst graph) CrsMatrix fillComplete                           0.001893 (1)
(nonconst graph) Kokkos sumInto                                   0.05297 (1)
(nonconst graph) sumIntoLocalValues: after first fillComplete     0.004121 (1)
(nonconst graph) sumIntoLocalValues: before first fillComplete    1.726 (1)
CrsGraph constructor                                              4.053e-06 (1)
CrsGraph::insertLocalIndices loop                                 0.003695 (1)
CrsMatrix constructor                                             3.6e-05 (1)
---------------------------------------------------------------------------------------------
CrsMatrix::insertLocalValues loop                                 0.00583 (1)
Whole benchmark                                                   1.852 (1)
createRowAndColMap                                                7.2e-05 (1)
=============================================================================================

lclNumInds: 10000
maxNumEntPerRow: 28
numToInsert: 27
numTrials: 100
resumeFill: true
=============================================================================================

                            TimeMonitor results over 1 processor

Timer Name                                                        Global time (num calls)
---------------------------------------------------------------------------------------------
(const graph) CrsMatrix fillComplete                              1.216e-05 (1)
(const graph) CrsMatrix resumeFill                                9.537e-07 (1)
(const graph) Kokkos sumInto                                      0.0526 (1)
(const graph) sumIntoLocalValues: after first fillComplete        1.776 (1)
(nonconst graph) CrsMatrix fillComplete                           0.001971 (1)
(nonconst graph) CrsMatrix resumeFill                             4.053e-06 (1)
(nonconst graph) Kokkos sumInto                                   0.0526 (1)
(nonconst graph) sumIntoLocalValues: after first fillComplete     1.753 (1)
(nonconst graph) sumIntoLocalValues: before first fillComplete    1.754 (1)
CrsGraph constructor                                              5.007e-06 (1)
---------------------------------------------------------------------------------------------
CrsGraph::insertLocalIndices loop                                 0.003651 (1)
CrsMatrix constructor                                             3.719e-05 (1)
CrsMatrix::insertLocalValues loop                                 0.005914 (1)
Whole benchmark                                                   5.401 (1)
createRowAndColMap                                                7.606e-05 (1)
=============================================================================================
```

mhoemmen commented:
Hold on, there might be an issue with the benchmark.  Stay tuned :-)

mhoemmen commented:
Oh, oops, sumIntoLocalValues reports an error (by returning -1) if the matrix is not fill active.  It returns -1 pretty quickly, which might explain the difference in times ;-).

mhoemmen commented:
I just fixed the benchmark issue.  Now the times are uniformly slow.

mhoemmen commented:
I went into `CrsMatrix::sumIntoLocalValuesImpl` and commented out the whole `if (graph.getProfileType() == StaticProfile)` test + its body.  Here are the results:
```
lclNumInds: 10000
maxNumEntPerRow: 28
numToInsert: 27
numTrials: 100
=============================================================================================

                            TimeMonitor results over 1 processor

Timer Name                                                        Global time (num calls)
---------------------------------------------------------------------------------------------
(const graph) Kokkos sumInto                                      0.06686 (1)
(const graph) Kokkos sumInto: after getLocalMatrix                0.06685 (1)
(const graph) sumIntoLocalValues: after first fillComplete        0.0969 (1)
(nonconst graph) Kokkos sumInto                                   0.06689 (1)
(nonconst graph) Kokkos sumInto: after getLocalMatrix             0.06687 (1)
(nonconst graph) sumIntoLocalValues: after first fillComplete     0.09689 (1)
(nonconst graph) sumIntoLocalValues: before first fillComplete    0.09677 (1)
Whole benchmark                                                   0.4393 (1)
=============================================================================================
```

With original sumIntoLocalValuesImpl implementation (including that if + its body):

```
lclNumInds: 10000
maxNumEntPerRow: 28
numToInsert: 27
numTrials: 100
=============================================================================================

                            TimeMonitor results over 1 processor

Timer Name                                                        Global time (num calls)
---------------------------------------------------------------------------------------------
(const graph) Kokkos sumInto                                      0.06682 (1)
(const graph) Kokkos sumInto: after getLocalMatrix                0.06682 (1)
(const graph) sumIntoLocalValues: after first fillComplete        1.71 (1)
(nonconst graph) Kokkos sumInto                                   0.06688 (1)
(nonconst graph) Kokkos sumInto: after getLocalMatrix             0.06686 (1)
(nonconst graph) sumIntoLocalValues: after first fillComplete     1.716 (1)
(nonconst graph) sumIntoLocalValues: before first fillComplete    1.717 (1)
Whole benchmark                                                   5.29 (1)
=============================================================================================
```

mhoemmen commented:
The code should still be correct with that `if` test + body commented out.  Also, the only part of the code below it that has anything to do with `DynamicProfile` is in `graph.get{Local,Global}KokkosRowView(rowInfo)`.  That method has a case for DynamicProfile, but it has other cases too; it's not exclusive to DynamicProfile.

tjfulle commented:
Wut??!!

tjfulle commented:
I coulda sworn the code after the block you commented out was specific to dynamic profile.

mhoemmen commented:
@tjfulle I've pushed the benchmark; please feel free to try it.  The default options should be fine.

mhoemmen commented:
If I hard-code the `findRelOffset` implementation directly into `sumIntoLocalValues` and bypass `getRowInfo` and other such calls, then `sumIntoLocalValues` is nearly as fast as `KokkosSparse::CrsMatrix::sumIntoValues`:
```
lclNumInds: 10000
maxNumEntPerRow: 28
numToInsert: 27
numTrials: 1000
=============================================================================================

                            TimeMonitor results over 1 processor

Timer Name                                                        Global time (num calls)
---------------------------------------------------------------------------------------------
(const graph) Kokkos sumInto                                      0.6684 (1)
(const graph) Kokkos sumInto: after getLocalMatrix                0.6684 (1)
(const graph) sumIntoLocalValues: after first fillComplete        0.8016 (1)
(nonconst graph) Kokkos sumInto                                   0.7116 (1)
(nonconst graph) Kokkos sumInto: after getLocalMatrix             0.7116 (1)
(nonconst graph) sumIntoLocalValues: after first fillComplete     0.8018 (1)
(nonconst graph) sumIntoLocalValues: before first fillComplete    0.7956 (1)
Whole benchmark                                                   3.795 (1)
=============================================================================================
```

mhoemmen commented:
The point of my last comment is to show that getting rid of DynamicProfile should actually improve performance, by getting rid of the need for abstractions and special cases inside `sumIntoLocalValues`.  We've seen this issue before, that abstractions like `getRowInfo`, despite looking pretty harmless, somehow manage to slow the code down.

trilinos-autotester commented:
Status Flag 'Pre-Test Inspection' - Auto Inspected - Inspection Is Not Necessary for this Pull Request.

trilinos-autotester commented:
Status Flag 'Pull Request AutoTester' - Testing Jenkins Projects:

<details>
<summary>
  <b>Pull Request Auto Testing STARTING</b> (click to expand)
</summary>

<h2>Build Information</h2>
<h3>Test Name: Trilinos_pullrequest_gcc_4.8.4</h3>
<ul>
  <li>Build Num: 3938</li>
  <li>Status: STARTED</li>
</ul>

<h4>Jenkins Parameters</h4>
<table>
  <tr>
    <th>Parameter Name</th>
    <th>Value</th>
  </tr>

  <tr>
    <td>COMPILER_MODULE</td>
    <td>sems-gcc/4.8.4</td>
  </tr>

  <tr>
    <td>JENKINS_BUILD_TYPE</td>
    <td>Release</td>
  </tr>

  <tr>
    <td>JENKINS_COMM_TYPE</td>
    <td>MPI</td>
  </tr>

  <tr>
    <td>JENKINS_DO_COMPLEX</td>
    <td>OFF</td>
  </tr>

  <tr>
    <td>JENKINS_JOB_TYPE</td>
    <td>Experimental</td>
  </tr>

  <tr>
    <td>MPI_MODULE</td>
    <td>sems-openmpi/1.8.7</td>
  </tr>

  <tr>
    <td>PULLREQUESTNUM</td>
    <td>5403</td>
  </tr>

  <tr>
    <td>TEST_REPO_ALIAS</td>
    <td>TRILINOS</td>
  </tr>

  <tr>
    <td>TRILINOS_SOURCE_BRANCH</td>
    <td>Tpetra-CrsMatrix-sumInto-perf</td>
  </tr>

  <tr>
    <td>TRILINOS_SOURCE_REPO</td>
    <td>https://github.com/mhoemmen/Trilinos</td>
  </tr>

  <tr>
    <td>TRILINOS_SOURCE_SHA</td>
    <td>912fba4da1acca1bbc95d7aaab2f25adee69cd64</td>
  </tr>

  <tr>
    <td>TRILINOS_TARGET_BRANCH</td>
    <td>develop</td>
  </tr>

  <tr>
    <td>TRILINOS_TARGET_REPO</td>
    <td>https://github.com/trilinos/Trilinos</td>
  </tr>

  <tr>
    <td>TRILINOS_TARGET_SHA</td>
    <td>c80327ff3efb396b208cfc7e51fd1c527408a54e</td>
  </tr>
</table>

<h2>Build Information</h2>
<h3>Test Name: Trilinos_pullrequest_intel_17.0.1</h3>
<ul>
  <li>Build Num: 3754</li>
  <li>Status: STARTED</li>
</ul>

<h4>Jenkins Parameters</h4>
<table>
  <tr>
    <th>Parameter Name</th>
    <th>Value</th>
  </tr>

  <tr>
    <td>PULLREQUESTNUM</td>
    <td>5403</td>
  </tr>

  <tr>
    <td>TEST_REPO_ALIAS</td>
    <td>TRILINOS</td>
  </tr>

  <tr>
    <td>TRILINOS_SOURCE_BRANCH</td>
    <td>Tpetra-CrsMatrix-sumInto-perf</td>
  </tr>

  <tr>
    <td>TRILINOS_SOURCE_REPO</td>
    <td>https://github.com/mhoemmen/Trilinos</td>
  </tr>

  <tr>
    <td>TRILINOS_SOURCE_SHA</td>
    <td>912fba4da1acca1bbc95d7aaab2f25adee69cd64</td>
  </tr>

  <tr>
    <td>TRILINOS_TARGET_BRANCH</td>
    <td>develop</td>
  </tr>

  <tr>
    <td>TRILINOS_TARGET_REPO</td>
    <td>https://github.com/trilinos/Trilinos</td>
  </tr>

  <tr>
    <td>TRILINOS_TARGET_SHA</td>
    <td>c80327ff3efb396b208cfc7e51fd1c527408a54e</td>
  </tr>
</table>

<h2>Build Information</h2>
<h3>Test Name: Trilinos_pullrequest_gcc_4.9.3_SERIAL</h3>
<ul>
  <li>Build Num: 2199</li>
  <li>Status: STARTED</li>
</ul>

<h4>Jenkins Parameters</h4>
<table>
  <tr>
    <th>Parameter Name</th>
    <th>Value</th>
  </tr>

  <tr>
    <td>PULLREQUESTNUM</td>
    <td>5403</td>
  </tr>

  <tr>
    <td>TEST_REPO_ALIAS</td>
    <td>TRILINOS</td>
  </tr>

  <tr>
    <td>TRILINOS_SOURCE_BRANCH</td>
    <td>Tpetra-CrsMatrix-sumInto-perf</td>
  </tr>

  <tr>
    <td>TRILINOS_SOURCE_REPO</td>
    <td>https://github.com/mhoemmen/Trilinos</td>
  </tr>

  <tr>
    <td>TRILINOS_SOURCE_SHA</td>
    <td>912fba4da1acca1bbc95d7aaab2f25adee69cd64</td>
  </tr>

  <tr>
    <td>TRILINOS_TARGET_BRANCH</td>
    <td>develop</td>
  </tr>

  <tr>
    <td>TRILINOS_TARGET_REPO</td>
    <td>https://github.com/trilinos/Trilinos</td>
  </tr>

  <tr>
    <td>TRILINOS_TARGET_SHA</td>
    <td>c80327ff3efb396b208cfc7e51fd1c527408a54e</td>
  </tr>
</table>

<h2>Build Information</h2>
<h3>Test Name: Trilinos_pullrequest_gcc_7.2.0</h3>
<ul>
  <li>Build Num: 1982</li>
  <li>Status: STARTED</li>
</ul>

<h4>Jenkins Parameters</h4>
<table>
  <tr>
    <th>Parameter Name</th>
    <th>Value</th>
  </tr>

  <tr>
    <td>PULLREQUESTNUM</td>
    <td>5403</td>
  </tr>

  <tr>
    <td>TEST_REPO_ALIAS</td>
    <td>TRILINOS</td>
  </tr>

  <tr>
    <td>TRILINOS_SOURCE_BRANCH</td>
    <td>Tpetra-CrsMatrix-sumInto-perf</td>
  </tr>

  <tr>
    <td>TRILINOS_SOURCE_REPO</td>
    <td>https://github.com/mhoemmen/Trilinos</td>
  </tr>

  <tr>
    <td>TRILINOS_SOURCE_SHA</td>
    <td>912fba4da1acca1bbc95d7aaab2f25adee69cd64</td>
  </tr>

  <tr>
    <td>TRILINOS_TARGET_BRANCH</td>
    <td>develop</td>
  </tr>

  <tr>
    <td>TRILINOS_TARGET_REPO</td>
    <td>https://github.com/trilinos/Trilinos</td>
  </tr>

  <tr>
    <td>TRILINOS_TARGET_SHA</td>
    <td>c80327ff3efb396b208cfc7e51fd1c527408a54e</td>
  </tr>
</table>

<h2>Build Information</h2>
<h3>Test Name: Trilinos_pullrequest_cuda_9.2</h3>
<ul>
  <li>Build Num: 1603</li>
  <li>Status: STARTED</li>
</ul>

<h4>Jenkins Parameters</h4>
<table>
  <tr>
    <th>Parameter Name</th>
    <th>Value</th>
  </tr>

  <tr>
    <td>JENKINS_JOB_TYPE</td>
    <td>Experimental</td>
  </tr>

  <tr>
    <td>PULLREQUESTNUM</td>
    <td>5403</td>
  </tr>

  <tr>
    <td>TEST_REPO_ALIAS</td>
    <td>TRILINOS</td>
  </tr>

  <tr>
    <td>TRILINOS_SOURCE_BRANCH</td>
    <td>Tpetra-CrsMatrix-sumInto-perf</td>
  </tr>

  <tr>
    <td>TRILINOS_SOURCE_REPO</td>
    <td>https://github.com/mhoemmen/Trilinos</td>
  </tr>

  <tr>
    <td>TRILINOS_SOURCE_SHA</td>
    <td>912fba4da1acca1bbc95d7aaab2f25adee69cd64</td>
  </tr>

  <tr>
    <td>TRILINOS_TARGET_BRANCH</td>
    <td>develop</td>
  </tr>

  <tr>
    <td>TRILINOS_TARGET_REPO</td>
    <td>https://github.com/trilinos/Trilinos</td>
  </tr>

  <tr>
    <td>TRILINOS_TARGET_SHA</td>
    <td>c80327ff3efb396b208cfc7e51fd1c527408a54e</td>
  </tr>
</table>

<h2>Build Information</h2>
<h3>Test Name: Trilinos_pullrequest_python_2</h3>
<ul>
  <li>Build Num: 112</li>
  <li>Status: STARTED</li>
</ul>

<h4>Jenkins Parameters</h4>
<table>
  <tr>
    <th>Parameter Name</th>
    <th>Value</th>
  </tr>

  <tr>
    <td>PULLREQUESTNUM</td>
    <td>5403</td>
  </tr>

  <tr>
    <td>TEST_REPO_ALIAS</td>
    <td>TRILINOS</td>
  </tr>

  <tr>
    <td>TRILINOS_SOURCE_BRANCH</td>
    <td>Tpetra-CrsMatrix-sumInto-perf</td>
  </tr>

  <tr>
    <td>TRILINOS_SOURCE_REPO</td>
    <td>https://github.com/mhoemmen/Trilinos</td>
  </tr>

  <tr>
    <td>TRILINOS_SOURCE_SHA</td>
    <td>912fba4da1acca1bbc95d7aaab2f25adee69cd64</td>
  </tr>

  <tr>
    <td>TRILINOS_TARGET_BRANCH</td>
    <td>develop</td>
  </tr>

  <tr>
    <td>TRILINOS_TARGET_REPO</td>
    <td>https://github.com/trilinos/Trilinos</td>
  </tr>

  <tr>
    <td>TRILINOS_TARGET_SHA</td>
    <td>c80327ff3efb396b208cfc7e51fd1c527408a54e</td>
  </tr>
</table>

<h2>Build Information</h2>
<h3>Test Name: Trilinos_pullrequest_python_3</h3>
<ul>
  <li>Build Num: 112</li>
  <li>Status: STARTED</li>
</ul>

<h4>Jenkins Parameters</h4>
<table>
  <tr>
    <th>Parameter Name</th>
    <th>Value</th>
  </tr>

  <tr>
    <td>PULLREQUESTNUM</td>
    <td>5403</td>
  </tr>

  <tr>
    <td>TEST_REPO_ALIAS</td>
    <td>TRILINOS</td>
  </tr>

  <tr>
    <td>TRILINOS_SOURCE_BRANCH</td>
    <td>Tpetra-CrsMatrix-sumInto-perf</td>
  </tr>

  <tr>
    <td>TRILINOS_SOURCE_REPO</td>
    <td>https://github.com/mhoemmen/Trilinos</td>
  </tr>

  <tr>
    <td>TRILINOS_SOURCE_SHA</td>
    <td>912fba4da1acca1bbc95d7aaab2f25adee69cd64</td>
  </tr>

  <tr>
    <td>TRILINOS_TARGET_BRANCH</td>
    <td>develop</td>
  </tr>

  <tr>
    <td>TRILINOS_TARGET_REPO</td>
    <td>https://github.com/trilinos/Trilinos</td>
  </tr>

  <tr>
    <td>TRILINOS_TARGET_SHA</td>
    <td>c80327ff3efb396b208cfc7e51fd1c527408a54e</td>
  </tr>
</table>

<h2>Using Repos:</h2>

<dl>
  <dt>Repo: TRILINOS (mhoemmen/Trilinos)</dt>
  <ul>
    <li>Branch: Tpetra-CrsMatrix-sumInto-perf</li>
    <li>SHA: 912fba4da1acca1bbc95d7aaab2f25adee69cd64</li>
    <li>Mode: TEST_REPO</li>
  </ul>
</dl>

Pull Request Author: mhoemmen
</details>

trilinos-autotester commented:
Status Flag 'Pull Request AutoTester' - Jenkins Testing: all Jobs PASSED

<details>
<summary>
  <b>Pull Request Auto Testing has PASSED</b> (click to expand)
</summary>

<h2>Build Information</h2>
<h3>Test Name: Trilinos_pullrequest_gcc_4.8.4</h3>
<ul>
  <li>Build Num: 3938</li>
  <li>Status: PASSED</li>
</ul>

<h4>Jenkins Parameters</h4>
<table>
  <tr>
    <th>Parameter Name</th>
    <th>Value</th>
  </tr>

  <tr>
    <td>COMPILER_MODULE</td>
    <td>sems-gcc/4.8.4</td>
  </tr>

  <tr>
    <td>JENKINS_BUILD_TYPE</td>
    <td>Release</td>
  </tr>

  <tr>
    <td>JENKINS_COMM_TYPE</td>
    <td>MPI</td>
  </tr>

  <tr>
    <td>JENKINS_DO_COMPLEX</td>
    <td>OFF</td>
  </tr>

  <tr>
    <td>JENKINS_JOB_TYPE</td>
    <td>Experimental</td>
  </tr>

  <tr>
    <td>MPI_MODULE</td>
    <td>sems-openmpi/1.8.7</td>
  </tr>

  <tr>
    <td>PULLREQUESTNUM</td>
    <td>5403</td>
  </tr>

  <tr>
    <td>TEST_REPO_ALIAS</td>
    <td>TRILINOS</td>
  </tr>

  <tr>
    <td>TRILINOS_SOURCE_BRANCH</td>
    <td>Tpetra-CrsMatrix-sumInto-perf</td>
  </tr>

  <tr>
    <td>TRILINOS_SOURCE_REPO</td>
    <td>https://github.com/mhoemmen/Trilinos</td>
  </tr>

  <tr>
    <td>TRILINOS_SOURCE_SHA</td>
    <td>912fba4da1acca1bbc95d7aaab2f25adee69cd64</td>
  </tr>

  <tr>
    <td>TRILINOS_TARGET_BRANCH</td>
    <td>develop</td>
  </tr>

  <tr>
    <td>TRILINOS_TARGET_REPO</td>
    <td>https://github.com/trilinos/Trilinos</td>
  </tr>

  <tr>
    <td>TRILINOS_TARGET_SHA</td>
    <td>c80327ff3efb396b208cfc7e51fd1c527408a54e</td>
  </tr>
</table>

<h2>Build Information</h2>
<h3>Test Name: Trilinos_pullrequest_intel_17.0.1</h3>
<ul>
  <li>Build Num: 3754</li>
  <li>Status: PASSED</li>
</ul>

<h4>Jenkins Parameters</h4>
<table>
  <tr>
    <th>Parameter Name</th>
    <th>Value</th>
  </tr>

  <tr>
    <td>PULLREQUESTNUM</td>
    <td>5403</td>
  </tr>

  <tr>
    <td>TEST_REPO_ALIAS</td>
    <td>TRILINOS</td>
  </tr>

  <tr>
    <td>TRILINOS_SOURCE_BRANCH</td>
    <td>Tpetra-CrsMatrix-sumInto-perf</td>
  </tr>

  <tr>
    <td>TRILINOS_SOURCE_REPO</td>
    <td>https://github.com/mhoemmen/Trilinos</td>
  </tr>

  <tr>
    <td>TRILINOS_SOURCE_SHA</td>
    <td>912fba4da1acca1bbc95d7aaab2f25adee69cd64</td>
  </tr>

  <tr>
    <td>TRILINOS_TARGET_BRANCH</td>
    <td>develop</td>
  </tr>

  <tr>
    <td>TRILINOS_TARGET_REPO</td>
    <td>https://github.com/trilinos/Trilinos</td>
  </tr>

  <tr>
    <td>TRILINOS_TARGET_SHA</td>
    <td>c80327ff3efb396b208cfc7e51fd1c527408a54e</td>
  </tr>
</table>

<h2>Build Information</h2>
<h3>Test Name: Trilinos_pullrequest_gcc_4.9.3_SERIAL</h3>
<ul>
  <li>Build Num: 2199</li>
  <li>Status: PASSED</li>
</ul>

<h4>Jenkins Parameters</h4>
<table>
  <tr>
    <th>Parameter Name</th>
    <th>Value</th>
  </tr>

  <tr>
    <td>PULLREQUESTNUM</td>
    <td>5403</td>
  </tr>

  <tr>
    <td>TEST_REPO_ALIAS</td>
    <td>TRILINOS</td>
  </tr>

  <tr>
    <td>TRILINOS_SOURCE_BRANCH</td>
    <td>Tpetra-CrsMatrix-sumInto-perf</td>
  </tr>

  <tr>
    <td>TRILINOS_SOURCE_REPO</td>
    <td>https://github.com/mhoemmen/Trilinos</td>
  </tr>

  <tr>
    <td>TRILINOS_SOURCE_SHA</td>
    <td>912fba4da1acca1bbc95d7aaab2f25adee69cd64</td>
  </tr>

  <tr>
    <td>TRILINOS_TARGET_BRANCH</td>
    <td>develop</td>
  </tr>

  <tr>
    <td>TRILINOS_TARGET_REPO</td>
    <td>https://github.com/trilinos/Trilinos</td>
  </tr>

  <tr>
    <td>TRILINOS_TARGET_SHA</td>
    <td>c80327ff3efb396b208cfc7e51fd1c527408a54e</td>
  </tr>
</table>

<h2>Build Information</h2>
<h3>Test Name: Trilinos_pullrequest_gcc_7.2.0</h3>
<ul>
  <li>Build Num: 1982</li>
  <li>Status: PASSED</li>
</ul>

<h4>Jenkins Parameters</h4>
<table>
  <tr>
    <th>Parameter Name</th>
    <th>Value</th>
  </tr>

  <tr>
    <td>PULLREQUESTNUM</td>
    <td>5403</td>
  </tr>

  <tr>
    <td>TEST_REPO_ALIAS</td>
    <td>TRILINOS</td>
  </tr>

  <tr>
    <td>TRILINOS_SOURCE_BRANCH</td>
    <td>Tpetra-CrsMatrix-sumInto-perf</td>
  </tr>

  <tr>
    <td>TRILINOS_SOURCE_REPO</td>
    <td>https://github.com/mhoemmen/Trilinos</td>
  </tr>

  <tr>
    <td>TRILINOS_SOURCE_SHA</td>
    <td>912fba4da1acca1bbc95d7aaab2f25adee69cd64</td>
  </tr>

  <tr>
    <td>TRILINOS_TARGET_BRANCH</td>
    <td>develop</td>
  </tr>

  <tr>
    <td>TRILINOS_TARGET_REPO</td>
    <td>https://github.com/trilinos/Trilinos</td>
  </tr>

  <tr>
    <td>TRILINOS_TARGET_SHA</td>
    <td>c80327ff3efb396b208cfc7e51fd1c527408a54e</td>
  </tr>
</table>

<h2>Build Information</h2>
<h3>Test Name: Trilinos_pullrequest_cuda_9.2</h3>
<ul>
  <li>Build Num: 1603</li>
  <li>Status: PASSED</li>
</ul>

<h4>Jenkins Parameters</h4>
<table>
  <tr>
    <th>Parameter Name</th>
    <th>Value</th>
  </tr>

  <tr>
    <td>JENKINS_JOB_TYPE</td>
    <td>Experimental</td>
  </tr>

  <tr>
    <td>PULLREQUESTNUM</td>
    <td>5403</td>
  </tr>

  <tr>
    <td>TEST_REPO_ALIAS</td>
    <td>TRILINOS</td>
  </tr>

  <tr>
    <td>TRILINOS_SOURCE_BRANCH</td>
    <td>Tpetra-CrsMatrix-sumInto-perf</td>
  </tr>

  <tr>
    <td>TRILINOS_SOURCE_REPO</td>
    <td>https://github.com/mhoemmen/Trilinos</td>
  </tr>

  <tr>
    <td>TRILINOS_SOURCE_SHA</td>
    <td>912fba4da1acca1bbc95d7aaab2f25adee69cd64</td>
  </tr>

  <tr>
    <td>TRILINOS_TARGET_BRANCH</td>
    <td>develop</td>
  </tr>

  <tr>
    <td>TRILINOS_TARGET_REPO</td>
    <td>https://github.com/trilinos/Trilinos</td>
  </tr>

  <tr>
    <td>TRILINOS_TARGET_SHA</td>
    <td>c80327ff3efb396b208cfc7e51fd1c527408a54e</td>
  </tr>
</table>

<h2>Build Information</h2>
<h3>Test Name: Trilinos_pullrequest_python_2</h3>
<ul>
  <li>Build Num: 112</li>
  <li>Status: PASSED</li>
</ul>

<h4>Jenkins Parameters</h4>
<table>
  <tr>
    <th>Parameter Name</th>
    <th>Value</th>
  </tr>

  <tr>
    <td>PULLREQUESTNUM</td>
    <td>5403</td>
  </tr>

  <tr>
    <td>TEST_REPO_ALIAS</td>
    <td>TRILINOS</td>
  </tr>

  <tr>
    <td>TRILINOS_SOURCE_BRANCH</td>
    <td>Tpetra-CrsMatrix-sumInto-perf</td>
  </tr>

  <tr>
    <td>TRILINOS_SOURCE_REPO</td>
    <td>https://github.com/mhoemmen/Trilinos</td>
  </tr>

  <tr>
    <td>TRILINOS_SOURCE_SHA</td>
    <td>912fba4da1acca1bbc95d7aaab2f25adee69cd64</td>
  </tr>

  <tr>
    <td>TRILINOS_TARGET_BRANCH</td>
    <td>develop</td>
  </tr>

  <tr>
    <td>TRILINOS_TARGET_REPO</td>
    <td>https://github.com/trilinos/Trilinos</td>
  </tr>

  <tr>
    <td>TRILINOS_TARGET_SHA</td>
    <td>c80327ff3efb396b208cfc7e51fd1c527408a54e</td>
  </tr>
</table>

<h2>Build Information</h2>
<h3>Test Name: Trilinos_pullrequest_python_3</h3>
<ul>
  <li>Build Num: 112</li>
  <li>Status: PASSED</li>
</ul>

<h4>Jenkins Parameters</h4>
<table>
  <tr>
    <th>Parameter Name</th>
    <th>Value</th>
  </tr>

  <tr>
    <td>PULLREQUESTNUM</td>
    <td>5403</td>
  </tr>

  <tr>
    <td>TEST_REPO_ALIAS</td>
    <td>TRILINOS</td>
  </tr>

  <tr>
    <td>TRILINOS_SOURCE_BRANCH</td>
    <td>Tpetra-CrsMatrix-sumInto-perf</td>
  </tr>

  <tr>
    <td>TRILINOS_SOURCE_REPO</td>
    <td>https://github.com/mhoemmen/Trilinos</td>
  </tr>

  <tr>
    <td>TRILINOS_SOURCE_SHA</td>
    <td>912fba4da1acca1bbc95d7aaab2f25adee69cd64</td>
  </tr>

  <tr>
    <td>TRILINOS_TARGET_BRANCH</td>
    <td>develop</td>
  </tr>

  <tr>
    <td>TRILINOS_TARGET_REPO</td>
    <td>https://github.com/trilinos/Trilinos</td>
  </tr>

  <tr>
    <td>TRILINOS_TARGET_SHA</td>
    <td>c80327ff3efb396b208cfc7e51fd1c527408a54e</td>
  </tr>
</table>
</details>


<br>[CDash Test Results for PR# 5403.](https://testing.sandia.gov/cdash/index.php?project=Trilinos&filtercombine=and&filtercombine=and&filtercombine=and&filtercombine=and&filtercombine=and&filtercombine=and&filtercount=2&showfilters=1&filtercombine=and&field1=buildname&compare1=63&value1=PR-5403-test&field2=buildstarttime&compare2=84&value2=NOW
)<br>

trilinos-autotester commented:
Status Flag 'Pre-Merge Inspection' - - This Pull Request Requires Inspection... The code must be inspected by a member of the Team before Testing/Merging
           WARNING: NO REVIEWERS HAVE BEEN REQUESTED FOR THIS PULL REQUEST!

trilinos-autotester commented:
All Jobs Finished; status = PASSED, However Inspection must be performed before merge can occur...

mhoemmen commented:
I think it's worth merging this one.  It's no longer a WIP.

trilinos-autotester commented:
Status Flag 'Pre-Merge Inspection' - - This Pull Request Requires Inspection... The code must be inspected by a member of the Team before Testing/Merging
           NO REVIEWS HAVE BEEN PERFORMED ON THIS PULL REQUEST!

trilinos-autotester commented:
All Jobs Finished; status = PASSED, However Inspection must be performed before merge can occur...

trilinos-autotester commented:
Status Flag 'Pre-Merge Inspection' - SUCCESS: The last commit to this Pull Request has been INSPECTED AND APPROVED by [ tjfulle ]!

trilinos-autotester commented:
Status Flag 'Pull Request AutoTester' - AutoMerge IS ENABLED, but the Label AT: AUTOMERGE is not set.  Either set Label AT: AUTOMERGE or manually merge the PR...

mhoemmen commented:
Thanks @tjfulle ! :-D
